server/src/main/java/org/elasticsearch/common/time/DateUtils.java

```diff
@@ -77,4 +78,224 @@ public static ZoneId of(String zoneId) {
         }
         return ZoneId.of(zoneId).normalized();
     }
+
+    /*
+     * begin of code that is partially copied from the joda time implementation in order to make calculations about utc rounding much
+     * faster than using java-time and assigning all those objects
+     *
+     */
+
+    private static final int DAYS_0000_TO_1970 = 719527;
+    private static final int MILLIS_PER_DAY = 86_400_000;
+    private static final long MILLIS_PER_YEAR = 31556952000L;
+
+    /**
+     * Rounds the given utc milliseconds sicne the epoch down to the next unit millis
+     *
+     * Note: This does not check for correctness of the result, as this only works with units smaller or equal than a day
+     *       In order to ensure the performane of this methods, there are no guards or checks in it
+     *
+     * @param utcMillis   the milliseconds since the epoch
+     * @param unitMillis  the unit to round to
+     * @return            the rounded milliseconds since the epoch
+     */
+    public static long roundFloor(long utcMillis, long unitMillis) {
+        if (utcMillis >= 0) {
+            return utcMillis - utcMillis % unitMillis;
+        } else {
+            utcMillis += 1;
+            return utcMillis - utcMillis % unitMillis - unitMillis;
+        }
+    }
+
+    /**
+     * Round down to the beginning of the quarter of the year of the specified time
+     * @param utcMillis the milliseconds since the epoch
+     * @return The milliseconds since the epoch rounded down to the quarter of the year
+     */
+    public static long roundQuarterOfYear(long utcMillis) {
+        int year = DateUtils.getYear(utcMillis);
+        int month = DateUtils.getMonthOfYear(utcMillis, year);
+        return DateUtils.of(year, Month.of(month).firstMonthOfQuarter().getValue());
```

Comment:
It looks like this constructs a `Month` object simply to round it down to a multiple of 3. I don't think it loses much clarity to do this by hand instead.